Mail Extender

askvortsov1 opened the pull request:
Improves result of #1986 
Fixes part of #1891 

**Changes proposed in this pull request:**
- Added Mail Driver extender (we want SES to be maintained by community, so let's make their lives easier)
- Integration tests for Mail Extender

**Reviewers should focus on:**
- What else should this extender do beyond just adding drivers? Maybe some parts of PrettyMail can be made easier via this?

**Confirmed**
- [x] Backend changes: tests are green (run `composer test`).

franzliedke commented on src/Mail/DriverInterface.php:
```diff
@@ -44,14 +46,12 @@ public function availableSettings(): array;
      */
-    // TODO: Uncomment for beta.13
-    //public function validate(SettingsRepositoryInterface $settings, Factory $validator): MessageBag;
+    public function validate(SettingsRepositoryInterface $settings, Factory $validator): MessageBag;
```
Oh, sorry, I hadn't seen this PR until now. I already took care of this part, so please rebase on the latest `master`.

askvortsov1 commented:
> * Calling `add` with a new driver makes the new driver appear in the `/api/mail-settings` endpoint
> * Calling `add` allows replacing existing drivers (e.g. a different set of fields appearing in `/api/mail-settings`)

Added! The tests are currently failing, but that's due to an underlying issue that I've fixed in https://github.com/flarum/core/pull/2036.

franzliedke commented on tests/integration/extenders/MailTest.php:
```diff
@@ -0,0 +1,116 @@
+use Swift_NullTransport;
+use Swift_Transport;
+
+class MailTest extends ApiControllerTestCase
```
Please extend from `Flarum\Tests\integration\TestCase` and use the `request()` method.

I consider the `ApiControllerTestCase` deprecated (I guess I should mention that) and have been slowly migrating tests over to testing real requests - to further decouple the tests from the endpoint's implementation (i.e. a controller).

franzliedke commented on src/Extend/Mail.php:
```diff
@@ -0,0 +1,32 @@
+{
+    protected $addDrivers = [];
+
+    public function add($humanName, $driver)
```
Please add doc blocks explaining the parameters. The human name isn't really a human name (it is currently shown in the dropdown, but that could change) - it's more like an identifier. The driver must be the name of a class (or binding) implementing the `MailDriverInterface`.

askvortsov1 commented:
Done, this now depends on https://github.com/flarum/core/pull/2052 to be merged first. We're going to need this authenticated request functionality anyway, so might as well get it done now!

franzliedke commented on src/Extend/Mail.php:
```diff
@@ -0,0 +1,38 @@
+     * @param string $identifier Identifier for mail driver. E.g. 'smtp' for SmtpDriver
+     * @param string $driver ::class attribute of driver class, which must implement Flarum\Mail\DriverInterface
+     */
+    public function add(string $identifier, $driver)
```
I am thinking this should either be `MailDriver::add` or `Mail::(add)Driver`. `Mail::add` feels a bit too generic.

What do you and others think?

askvortsov1 replied:
Agreed. I'd say either ::driver() or ::addDriver()

askvortsov1 replied:
I went with addDriver in case we want to add more mail extension functionality later. I'd prefer to keep the extenders themselves flexible

franzliedke commented on tests/integration/extenders/MailTest.php:
```diff
@@ -0,0 +1,100 @@
+use Swift_NullTransport;
+use Swift_Transport;
+
+class MailTest extends AuthenticatedTestCase
```
This base class does not exist anymore. Please rebase on the latest master and change accordingly.

askvortsov1 replied:
Sorry got sidetracked with my categories extension, I'll have all my extender PRs rebased sometime today

askvortsov1 replied:
Done!
